File: sharedModule/src/main/java/ru/nekit/android/qls/utils/AnimationUtils.java
package ru.nekit.android.qls.utils;


import android.animation.Animator;
import android.animation.ArgbEvaluator;
import android.animation.ObjectAnimator;
import android.animation.TimeInterpolator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.support.annotation.ColorInt;
import android.support.annotation.ColorRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.view.View;

import ru.nekit.android.shared.R;

public class AnimationUtils {

    public static ValueAnimator getColorAnimator(@NonNull Context context, @ColorRes int startColor,
                                                 @ColorRes int endColor, int duration,
                                                 @NonNull final View view,
                                                 @NonNull TimeInterpolator interpolator) {
        return getColorAnimator(ContextCompat.getColor(context, startColor),
                ContextCompat.getColor(context, endColor), duration, view, interpolator);
    }

    public static ValueAnimator getColorAnimator(@NonNull Context context, @ColorRes int startColor,
                                                 @ColorRes int endColor, int duration,
                                                 @NonNull final View view) {
        return getColorAnimator(ContextCompat.getColor(context, startColor),
                ContextCompat.getColor(context, endColor), duration, view, null);
    }

    public static ValueAnimator getColorAnimator(@ColorInt int startColor, @ColorInt int endColor,
                                                 int duration, @NonNull final View view) {
        return getColorAnimator(startColor, endColor, duration, view, null);
    }

    public static void shake(@NonNull View view) {
        ObjectAnimator rotate = ObjectAnimator.ofFloat(view, "rotation", 0f, 20f, 0f, -20f, 0f);
        rotate.setRepeatCount(5);
        rotate.setDuration(100);
        rotate.start();
    }

    public static ValueAnimator getColorAnimator(@ColorInt int startColor, @ColorInt int endColor,
                                                 int duration, @NonNull final View view,
                                                 @Nullable TimeInterpolator interpolator) {
        final ValueAnimator colorAnimation = ValueAnimator.ofObject(new ArgbEvaluator(),
                startColor, endColor);
        colorAnimation.setDuration(duration);
        if (interpolator != null) {
            colorAnimation.setInterpolator(interpolator);
        }
        colorAnimation.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                colorAnimation.removeAllListeners();
                colorAnimation.removeAllUpdateListeners();
            }

            @Override
            public void onAnimationCancel(Animator animation) {
            }

            @Override
            public void onAnimationRepeat(Animator animation) {

            }
        });
        colorAnimation.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animator) {
                view.setBackgroundColor((int) animator.getAnimatedValue());
            }
        });
        return colorAnimation;
    }

    public static void fadeAnimation(@NonNull Context context, @Nullable View view, boolean fadeOut) {
        if (view != null) {
            if (!fadeOut) {
                view.setAlpha(0);
            }
            view.animate().alpha(fadeOut ? 0 : 1).setDuration(
                    context.getResources().getInteger(R.integer.short_animation_duration));
        }
    }
}

File: sharedModule/src/main/java/ru/nekit/android/utils/AnimationUtils.kt
package ru.nekit.android.utils


import android.animation.*
import android.content.Context
import android.support.annotation.ColorInt
import android.support.annotation.ColorRes
import android.support.annotation.IntegerRes
import android.support.v4.content.ContextCompat
import android.support.v4.view.animation.FastOutLinearInInterpolator
import android.view.View
import android.view.ViewPropertyAnimator
import ru.nekit.android.shared.R

object AnimationUtils {

    fun getColorAnimator(context: Context, @ColorRes startColor: Int,
                         @ColorRes endColor: Int, duration: Int,
                         view: View,
                         interpolator: TimeInterpolator): ValueAnimator {
        return getColorAnimator(ContextCompat.getColor(context, startColor),
                ContextCompat.getColor(context, endColor), duration, view, interpolator)
    }

    fun getColorAnimator(context: Context, @ColorRes startColor: Int,
                         @ColorRes endColor: Int, duration: Int,
                         view: View): ValueAnimator {
        return getColorAnimator(ContextCompat.getColor(context, startColor),
                ContextCompat.getColor(context, endColor), duration, view, null)
    }

    fun shake(view: View) {
        ObjectAnimator.ofFloat(view, "rotation", 0f, 20f, 0f, -20f, 0f).apply {
            repeatCount = 5
            duration = 100
            start()
        }
    }

    @JvmOverloads
    fun getColorAnimator(@ColorInt startColor: Int, @ColorInt endColor: Int,
                         duration: Int, view: View,
                         interpolator: TimeInterpolator? = null): ValueAnimator {
        val colorAnimation = ValueAnimator.ofObject(ArgbEvaluator(),
                startColor, endColor)
        colorAnimation.duration = duration.toLong()
        if (interpolator != null) {
            colorAnimation.interpolator = interpolator
        }
        colorAnimation.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animation: Animator) {}

            override fun onAnimationEnd(animation: Animator) {
                colorAnimation.removeAllListeners()
                colorAnimation.removeAllUpdateListeners()
            }

            override fun onAnimationCancel(animation: Animator) {}

            override fun onAnimationRepeat(animation: Animator) {

            }
        })
        colorAnimation.addUpdateListener { animator -> view.setBackgroundColor(animator.animatedValue as Int) }
        return colorAnimation
    }

    fun fadeAnimation(view: View?, fadeOut: Boolean, duration: Long, actionEnd: () -> Unit) {
        if (view != null) {
            if (!fadeOut) {
                view.alpha = 0f
            }
            view.animate().alpha((if (fadeOut) 0 else 1).toFloat()).setDuration(duration).withEndAction(actionEnd)
        }
    }

    fun fadeAnimation(view: View?, fadeOut: Boolean, duration: Long) = fadeAnimation(view, fadeOut, duration) {}

    fun animate(view: View, duration: Long,
                interpolator: TimeInterpolator,
                startAnimate: (ViewPropertyAnimator) -> ViewPropertyAnimator,
                actionStart: () -> Unit,
                endAnimate: (ViewPropertyAnimator) -> ViewPropertyAnimator,
                actionEnd: () -> Unit): ViewPropertyAnimator =
            view.animate().apply {
                startAnimate(withLayer()
                        .setInterpolator(interpolator)
                        .setDuration(duration)
                        .withEndAction {
                            actionStart()
                            endAnimate(setDuration(duration)
                                    .withEndAction {
                                        view.setLayerType(View.LAYER_TYPE_NONE, null)
                                        actionEnd()
                                    })
                        }
                )
            }

    fun fadeOutAndIn(view: View, duration: Long, actionStart: () -> Unit, actionEnd: () -> Unit) =
            animate(view, duration, FastOutLinearInInterpolator(),
                    { it.alpha(0F) }, actionStart,
                    { it.alpha(1F) }, actionEnd)

    fun fadeOutAndIn(view: View, duration: Long, action: () -> Unit) = fadeOutAndIn(view, duration, action, {})

}

enum class Delay(@IntegerRes private val resourceId: Int) {

    LONG(R.integer.long_animation_duration),
    SHORT(R.integer.short_animation_duration),
    SMALL(R.integer.small_animation_duration),
    KEYBOARD(R.integer.keyboard_delay),
    VIBRATION(R.integer.vibration_delay);

    fun get(context: Context) = context.resources.getInteger(resourceId).toLong()
}